[Feat] add I2VGenXL for image-to-video generation

sayakpaul opened the pull request:
# What does this PR do?

Fixes: https://github.com/huggingface/diffusers/issues/6186. 

## Test code

```python
import torch
from diffusers import I2VGenXLPipeline
from diffusers.utils import load_image, export_to_gif

# repo is currently private to the diffusers team
repo_id = "diffusers/i2vgen-xl" # TODO change checkpoint path after move
pipeline = I2VGenXLPipeline.from_pretrained(repo_id, torch_dtype=torch.float16, variant="fp16").to("cuda")

image_url = "https://github.com/ali-vilab/i2vgen-xl/blob/main/data/test_images/img_0009.png?download=true"
image = load_image(image_url).convert("RGB")
prompt = "Papers were floating in the air on a table in the library"

generator = torch.manual_seed(8888)
frames = pipeline(
    prompt=prompt,
    image=image,
    generator=generator
).frames[0]

print(export_to_gif(frames))
```

For memory optimization, use `enable_model_cpu_offload()`. 


## TODO

- [x] Debug video quality
- [x] Docs
- [x] Tests
- [x] Eliminate `einops` dependency 
- [x] General cleanup

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_6665). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

sayakpaul commented:
@DN6 here are some more updates. 

Currently, the generated videos are noisy. A sample is available [here](https://huggingface.co/datasets/diffusers/docs-images/blob/main/i2vgen_xl_images/frog.gif). 

I tried to investigate the potential suspects: 

* Use the [original sinusoidal embedding implementation](https://github.com/ali-vilab/i2vgen-xl/blob/2539c9262ff8a2a22fa9daecbfd13f0a2dbc32d0/tools/modules/unet/util.py#L177C1-L189C13). Didn't help. 
* Use the same image latents when using classifier-free guidance (see the original block [here](https://github.com/ali-vilab/i2vgen-xl/blob/2539c9262ff8a2a22fa9daecbfd13f0a2dbc32d0/tools/inferences/inference_i2vgen_entrance.py#L194C1-L196C109), relevant diffusers block is [here](https://github.com/huggingface/diffusers/blob/11fd646932b58b4a068a9f272e65c653abc7e2f7/src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py#L411)). Didn't help. 

My inference script is as follows:

<details>
<summary>Inference script</summary>

```python
from diffusers.pipelines.i2vgen_xl.pipeline_i2vgen_xl import I2VGenXLPipeline
import torch
import PIL
from moviepy.editor import VideoFileClip
from diffusers import DDIMScheduler
from diffusers.utils import load_image, export_to_video
import argparse


folder = "i2vgen-xl-diffusers"
pipeline = I2VGenXLPipeline.from_pretrained(folder, torch_dtype=torch.float16).to(
    "cuda"
)
scheduler = DDIMScheduler.from_pretrained(
    "cerspense/zeroscope_v2_576w",
    subfolder="scheduler",
    clip_sample=False,
    timestep_spacing="linspace",
    beta_schedule="linear",
    steps_offset=1,
)
pipeline.scheduler = scheduler
negative_prompt = "Distorted, discontinuous, Ugly, blurry, low resolution, motionless, static, disfigured, disconnected limbs, Ugly faces, incomplete arms"


def generate_video(args):
    if "https" in args.image_url_or_path:
        image = load_image(args.image_url_or_path).convert("RGB")
    else:
        image = PIL.Image.open(args.image_url_or_path).convert("RGB")

    generator = torch.manual_seed(8888)

    frames = pipeline(
        prompt=args.prompt,
        image=image,
        negative_prompt=negative_prompt,
        generator=generator,
    ).frames
    video_path = export_to_video(frames, output_video_path=args.video_out_path, fps=16)
    print(f"Video available at: {video_path}")

    if args.render_gif:
        clip = VideoFileClip(video_path)
        gif_path = video_path.replace(".mp4", ".gif")
        clip.write_gif(gif_path, fps=clip.fps)
        print(f"GIF available at: {gif_path}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--image_url_or_path",
        type=str,
        default="https://huggingface.co/datasets/diffusers/docs-images/resolve/main/i2vgen_xl_images/img_0001.jpg",
    )
    parser.add_argument(
        "--prompt",
        type=str,
        default="A green frog floats on the surface of the water on green lotus leaves, with several pink lotus flowers, in a Chinese painting style.",
    )
    parser.add_argument("--video_out_path", type=str, default=None)
    parser.add_argument("--render_gif", action="store_true")
    args = parser.parse_args()

    generate_video(args)
```
</details>

If we use the scheduler as is, the resultant video looks gibberish (see [here](https://huggingface.co/datasets/diffusers/docs-images/blob/main/i2vgen_xl_images/frog_without_scheduler_changes.gif)). 

But I think it's in a good place for you to take over as I will be off from tomorrow till this week. 

I have listed some TODOs in the PR description as well as have made sufficient comments in the code for us to track the subsequent cleanups. 

As also mentioned, please do also consider the [Notion doc](https://www.notion.so/huggingface2/text-to-video-damo-vilab-i2vgen-xl-f8662ff793ab4e129dbaf74fc6eb8141) (internal) while you are working on this.  

Let me know if there's anything unclear.

sayakpaul commented on src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py:
```diff
@@ -0,0 +1,858 @@
+        output_type: Optional[str] = "pil",
+        return_dict: bool = True,
+        cross_attention_kwargs: Optional[Dict[str, Any]] = None,
+        clip_skip: Optional[int] = 1,
```
I think it's okay to set `clip_skip` to 1 here since this is default in the original implementation. Just like we are setting `guidance_scale` and `num_inference_steps` to the original default values.

sayakpaul commented on .github/workflows/pr_tests.yml:
```diff
@@ -113,7 +113,7 @@ jobs:
       if: ${{ matrix.config.framework == 'pytorch_examples' }}
       run: |
-        python -m pip install peft
+        python -m pip install peft==0.7.1
```
Check https://huggingface.slack.com/archives/C04L3MWLE6B/p1706601893641009

patrickvonplaten replied:
Could it make sense to do this is a seperate PR?

sayakpaul replied:
Yeah. Will be removed altogether.

sayakpaul commented:
The failing test seems to be completely unrelated and it fails on the `main` too.

patrickvonplaten commented on docs/source/en/api/pipelines/i2vgenxl.md:
```diff
@@ -0,0 +1,57 @@
+* It can generate videos of quality that is often on par with [Stable Video Diffusion](../../using-diffusers/svd) (SVD). 
+* Unlike SVD, it additionally accepts text prompts as inputs. 
+* It can generate higher resolution videos. 
+* When using the [`DDIMScheduler`] (which is default for this pipeline), less than 50 steps for inference lead to weird results.
```
```suggestion
* When using the [`DDIMScheduler`] (which is default for this pipeline), less than 50 steps for inference leads to bad results.
```

patrickvonplaten commented on src/diffusers/models/attention.py:
```diff
@@ -252,7 +252,7 @@ def __init__(
                 "layer_norm",
             )
-        elif not self.use_ada_layer_norm_single:
+        elif not self.use_ada_layer_norm_single and norm_type != "layer_norm_i2vgen":
```
```suggestion
        elif norm_type in ["layer_norm", "ada_norm", "ada_norm_zero"]:
        
```

Can we try to refactor the whole layer norm logic here? Instead of creating booleans, let's write if-else statements using types + string matching

sayakpaul replied:
Done in be518c87.

patrickvonplaten commented on src/diffusers/models/attention.py:
```diff
@@ -372,10 +372,12 @@ def forward(
-        elif not self.use_ada_layer_norm_single:
-            norm_hidden_states = self.norm3(hidden_states)
+        # i2vgen doesn't have this norm 🤷‍♂️
+        if hasattr(self, "norm3") and self.norm3 is not None:
```
Let's refactor this and all if-else statements that following to use `if self.norm_type in ["...."]:`

sayakpaul replied:
Done in be518c87.

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+    return inputs
+
+
+def _collapse_frames_into_batch(sample: torch.Tensor) -> torch.Tensor:
+    batch_size, channels, num_frames, height, width = sample.shape
+    sample = sample.permute(0, 2, 1, 3, 4).reshape(batch_size * num_frames, channels, height, width)
+
+    return sample
```
I think this function is only used twice and can be written more or less as a one-liner -> can we remove it and just write out the function instead?

sayakpaul replied:
I think the main reason why we did it this way is because the shapes of the `sample` is different for the two instances this method is being used. 

So, we have to unpack the shape into `batch_size, channels, num_frames, height, width` each time. So, made sense to delegate this utility.

patrickvonplaten replied:
This function is used twice and can be written as a one-liner:
```py
sample = sample.permute(0, 2, 1, 3, 4).reshape(sample.shape[0] * sample.shape[2], sample.shape[1], sample.shape[3], sample.shape[4])
```

=> better to not have this function

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+        return processors
+
+    # Copied from diffusers.models.unets.unet_2d_condition.UNet2DConditionModel.set_attention_slice
+    def set_attention_slice(self, slice_size: Union[str, int, List[int]]) -> None:
```
I don't think we should add `set_attention_slice` to the UNet anymore. No one uses attention slicing since flash attention anymore IMO

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+        )
+
+        # image embedding
+        self.image_latents_proj_in = nn.Sequential(
```
Generally not a big fan of `nn.Sequential` but ok

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+    def forward(
+        self,
+        sample: torch.FloatTensor,
+        timestep: Union[torch.Tensor, float, int],
```
IMO it could make more sense to force `timestep` to be a tensor here (make sure the pipeline is responsible for the dtype casting and not the unet. This would make the unet less error prone and more efficient

sayakpaul replied:
I think this warrants for a pipeline-wide PR. I can do this after this integration. That will be better, probably.

sayakpaul replied:
https://github.com/huggingface/diffusers/issues/6763

patrickvonplaten replied:
I think we should only update the `pipeline_i2vgen_xl.py` here - no need to update all the other pipelines (it's not possible anymore anyways due to BW). It's a bit of a pity because `timestep: Union[torch.Tensor, float, int]` will now go into a release

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+        sample: torch.FloatTensor,
+        timestep: Union[torch.Tensor, float, int],
+        fps: torch.Tensor,
+        image_latents: torch.Tensor,
+        image_embeddings: Optional[torch.Tensor] = None,
```
No docstring here - what are these tensors doing?

sayakpaul replied:
Done in 88f03a39.

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+            attention_mask = attention_mask.unsqueeze(1)
+
+        # 1. time
+        timesteps = timestep
+        timesteps = _to_tensor(timestep, sample.device)
```
```suggestion
        timesteps = _to_tensor(timestep, sample.device)
```

patrickvonplaten commented on src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py:
```diff
@@ -0,0 +1,889 @@
+                    latent_model_input,
+                    t,
+                    encoder_hidden_states=prompt_embeds,
+                    fps=fps_tensor,
```
`fps` is always a tensor here no? I think we don't need to run `_to_tensor` on the `fps` inside the unet then

sayakpaul replied:
Done in a9c23e8b.

patrickvonplaten commented on src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py:
```diff
@@ -0,0 +1,889 @@
+    return image
+
+
+def _center_crop_wide(
```
Can't we use the image_processor for this?

sayakpaul replied:
The resampling algorithm is different in the image processor which is exactly why I didn't use that.

yiyixuxu replied:
we can set it in the image processor config https://github.com/huggingface/diffusers/blob/8e9466350333aae8fbbff2c7b8da5f08778bbd02/src/diffusers/image_processor.py#L68

patrickvonplaten commented on src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py:
```diff
@@ -0,0 +1,889 @@
+# https://github.com/ali-vilab/i2vgen-xl/blob/main/utils/transforms.py.
+
+
+def _convert_pt_to_pil(image: Union[torch.Tensor, List[torch.Tensor]]):
```
Can't we use the `image_processor` for this?

sayakpaul replied:
Here's the sequence of the methods that need to be called with `VaeImageProcessor`:

* `pt_to_numpy()`
* `numpy_to_pil()`

I think we could do that. 

Done in: ec5694ad

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+        image_embeddings: Optional[torch.Tensor] = None,
+        encoder_hidden_states: Optional[torch.Tensor] = None,
+        timestep_cond: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
```
```suggestion
```

I don't think `attention_mask` is ever passed to the UNet

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+                if hasattr(upsample_block, k) or getattr(upsample_block, k, None) is not None:
+                    setattr(upsample_block, k, None)
+
+    # Copied from diffusers.models.unets.unet_2d_condition.UNet2DConditionModel.unload_lora
+    def unload_lora(self):
+        """Unloads LoRA weights."""
+        deprecate(
+            "unload_lora",
+            "0.28.0",
+            "Calling `unload_lora()` is deprecated and will be removed in a future version. Please install `peft` and then call `disable_adapters().",
+        )
+        for module in self.modules():
+            if hasattr(module, "set_lora_layer"):
+                module.set_lora_layer(None)
```
```suggestion
```

Let's not add deprecated code

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+            prev_output_channel = output_channel
+
+        # out
+        if norm_num_groups is not None:
+            self.conv_norm_out = nn.GroupNorm(
+                num_channels=block_out_channels[0], num_groups=norm_num_groups, eps=norm_eps
+            )
+            self.conv_act = get_activation("silu")
+        else:
+            self.conv_norm_out = None
+            self.conv_act = None
```
Do we need the if-else statements here -> let's try to make the unet as simple as possible and to not just copy the code here

sayakpaul replied:
Done in 5cbaf2e7.

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+
+        self.sample_size = sample_size
+
+        if num_attention_heads is not None:
+            raise NotImplementedError(
+                "At the moment it is not possible to define the number of attention heads via `num_attention_heads` because of a naming issue as described in https://github.com/huggingface/diffusers/issues/2011#issuecomment-1547958131. Passing `num_attention_heads` will only be supported in diffusers v0.19."
+            )
+
```
```suggestion
        if num_attention_heads is not None:
            raise NotImplementedError(
                "At the moment it is not possible to define the number of attention heads via `num_attention_heads` because of a naming issue as described in https://github.com/huggingface/diffusers/issues/2011#issuecomment-1547958131. Passing `num_attention_heads` will only be supported in diffusers v0.19."
            )

```

Can we correct the falsy naming here? Let's maybe try to not add `attention_head_dim` anymore to the init

sayakpaul replied:
Took care of it in 8c980df1.

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+        norm_num_groups: Optional[int] = 32,
+        norm_eps: float = 1e-5,
+        cross_attention_dim: int = 1024,
+        attention_head_dim: Union[int, Tuple[int]] = 64,
```
```suggestion
```

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+        norm_eps: float = 1e-5,
+        cross_attention_dim: int = 1024,
+        attention_head_dim: Union[int, Tuple[int]] = 64,
+        num_attention_heads: Optional[Union[int, Tuple[int]]] = None,
```
```suggestion
        num_attention_heads: Optional[Union[int, Tuple[int]]] = 64,
```

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+            "CrossAttnUpBlock3D",
+        ),
+        block_out_channels: Tuple[int, ...] = (320, 640, 1280, 1280),
+        layers_per_block: int = 2,
+        downsample_padding: int = 1,
+        mid_block_scale_factor: float = 1,
+        act_fn: str = "silu",
+        norm_num_groups: Optional[int] = 32,
+        norm_eps: float = 1e-5,
```
```suggestion
        layers_per_block: int = 2,
        downsample_padding: int = 1,
        mid_block_scale_factor: float = 1,
        act_fn: str = "silu",
        norm_num_groups: Optional[int] = 32,
        norm_eps: float = 1e-5,
```

Do we need all this configs?

sayakpaul replied:
Removed some of them in e6c07b56. 

Only kept the essential ones.

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+            logger.info("Forward upsample size to force interpolation output size.")
+            forward_upsample_size = True
+
+        # prepare attention_mask
+        if attention_mask is not None:
+            attention_mask = (1 - attention_mask.to(sample.dtype)) * -10000.0
+            attention_mask = attention_mask.unsqueeze(1)
+
```
```suggestion
```

sayakpaul replied:
Done in 5ba9b4ad.

patrickvonplaten commented on src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py:
```diff
@@ -0,0 +1,889 @@
+            else:
+                attention_mask = None
+
+            if clip_skip is None:
```
Do we need to add `clip_skip` right away? Is it used for i2vgen_xl?

sayakpaul replied:
Yeah. They use the penultimate layer for text representations. This is why the `clip_skip` value is set to 1 in the pipeline call.

patrickvonplaten commented on src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py:
```diff
@@ -0,0 +1,889 @@
+
+        if prompt_embeds is None:
+            # textual inversion: procecss multi-vector tokens if necessary
+            if isinstance(self, TextualInversionLoaderMixin):
```
Is textual inversion used for i2vgen_xl?

sayakpaul replied:
Removed in a682dca0.

patrickvonplaten commented on src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py:
```diff
@@ -0,0 +1,889 @@
+        """
+        # set lora scale so that monkey patched LoRA
+        # function of text encoder can correctly access it
+        if lora_scale is not None and isinstance(self, LoraLoaderMixin):
```
Are LoRAs used for I2vgen_xl already?

sayakpaul replied:
No, thought it would be best to just have the support anyway from day 0.

patrickvonplaten replied:
Generally, I'd try to only add features that are used to not bloat the code

patrickvonplaten reviewed (commented):
Design-wise I think this PR goes into the right direction! Let's maybe just make sure that the code we add is as trimmed as possible. All configs, options that are not used shouldn't be added

sayakpaul commented on .github/workflows/pr_tests.yml:
```diff
@@ -113,7 +113,7 @@ jobs:
       if: ${{ matrix.config.framework == 'pytorch_examples' }}
       run: |
-        python -m pip install peft
+        python -m pip install peft==0.7.1
```
```suggestion
        python -m pip install peft
```

DN6 commented:
@sayakpaul I think the norm changes to `BasicTransformerBlock` can be reverted since we're using `I2VGenTransformerInBlock` which is simpler. Forgot to make that change.

sayakpaul commented:
@patrickvonplaten I have addressed your comments, WDYT? 

@DN6 
> @sayakpaul I think the norm changes to `BasicTransformerBlock` can be reverted since we're using `I2VGenTransformerInBlock` which is simpler. Forgot to make that change.

I don't see any `I2VGenTransformerInBlock` in the code.

Also, since now the configuration variables have reduced now, we'd need to run the conversion again. I think it's best to do it after an approval. What do you think?

yiyixuxu commented on src/diffusers/models/attention.py:
```diff
@@ -182,11 +178,11 @@ def __init__(
         # Define 3 blocks. Each block has its own normalization layer.
         # 1. Self-Attn
-        if self.use_ada_layer_norm:
+        if norm_type == "ada_norm":
```
nice :)

yiyixuxu commented on src/diffusers/models/attention.py:
```diff
@@ -252,7 +248,8 @@ def __init__(
-        elif not self.use_ada_layer_norm_single:
+
+        elif norm_type in ["ada_norm_zero", "ada_norm", "layer_norm", "ada_norm_continuous"]:
             self.norm3 = nn.LayerNorm(dim, norm_eps, norm_elementwise_affine)
```
I think it is better to still have `norm3` but set to `None`
```suggestion
            self.norm3 = nn.LayerNorm(dim, norm_eps, norm_elementwise_affine)
     elif norm_type == "layer_norm_i2vgen":
            self.norm3 = None
```

sayakpaul replied:
Sweet!

yiyixuxu commented on src/diffusers/models/attention.py:
```diff
@@ -332,9 +329,9 @@ def forward(
             **cross_attention_kwargs,
         )
-        if self.use_ada_layer_norm_zero:
+        if self.num_embeds_ada_norm is not None and self.norm_type == "ada_norm_zero":
```
I think it is guaranteed  `self.num_embeds_ada_norm` is not `None` here - it will get a value error during `__init__`
```suggestion
        if self.norm_type == "ada_norm_zero":
```

yiyixuxu commented on src/diffusers/models/attention.py:
```diff
@@ -347,20 +344,22 @@ def forward(
         # 3. Cross-Attention
         if self.attn2 is not None:
-            if self.use_ada_layer_norm:
+            if (self.num_embeds_ada_norm is not None) and self.norm_type == "ada_norm":
```
```suggestion
            if self.norm_type == "ada_norm":
```

yiyixuxu commented on src/diffusers/models/attention.py:
```diff
@@ -347,20 +344,22 @@ def forward(
+            if (self.num_embeds_ada_norm is not None) and self.norm_type == "ada_norm":
                 norm_hidden_states = self.norm2(hidden_states, timestep)
-            elif self.use_ada_layer_norm_zero or self.use_layer_norm:
+            elif ((self.num_embeds_ada_norm is not None) and self.norm_type == "ada_norm_zero") or (
+                self.norm_type in ["layer_norm", "layer_norm_i2vgen"]
+            ):
```
```suggestion
            elif self.norm_type in [""ada_norm_zero", layer_norm", "layer_norm_i2vgen"]
            ):
```

sayakpaul replied:
```suggestion
            elif self.norm_type in ["ada_norm_zero", layer_norm", "layer_norm_i2vgen"]
            ):
```

yiyixuxu commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,685 @@
+        for name, module in self.named_children():
+            fn_recursive_attn_processor(name, module, processor)
+
+    def enable_forward_chunking(self, chunk_size: Optional[int] = None, dim: int = 0) -> None:
```
can we add a `#copied from` here

yiyixuxu commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,685 @@
+        for module in self.children():
+            fn_recursive_feed_forward(module, chunk_size, dim)
+
+    def disable_forward_chunking(self):
```
copied from ?

yiyixuxu commented on src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py:
```diff
@@ -0,0 +1,880 @@
+        )
+        self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1)
+        # `do_resize=False` as we do custom resizing.
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor, do_resize=False)
```
feel free to add your custom resizing method to image_processor https://github.com/huggingface/diffusers/blob/8e9466350333aae8fbbff2c7b8da5f08778bbd02/src/diffusers/image_processor.py#L333

sayakpaul replied:
So, the supported resizing methods are different from the one being used in the pipeline. Hence didn't have any other option.

yiyixuxu replied:
we could add the resize method to the image processor and then add a new option here for `resize_mode` - but I agree with you that it makes sense to have its own image processor for videos so this can be refactored in a future PR I guess :) !

patrickvonplaten replied:
Let's also not forgot to refactor this please

sayakpaul replied:
What refactoring here? Are you referring to the separate video processing class we are brainstorming?

yiyixuxu commented on src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py:
```diff
@@ -0,0 +1,880 @@
+
+        return image_latents
+
+    def prepare_latents(
```
is this copied from some where?

yiyixuxu commented on src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py:
```diff
@@ -0,0 +1,880 @@
+
+        # 3.2 Encode image prompt
+        # 3.2.1 Image encodings.
+        image_embeddings = self._encode_image(image, width, device, num_videos_per_prompt)
+
+        # 3.2.2 Image latents.
+        resized_image = _center_crop_wide(image, (width, height))
```
then we don't do to `_center_crop_wide` inside `_encode_image`
```suggestion
        resized_image = _center_crop_wide(image, (width, height))
        image_embeddings = self._encode_image(resized_image, width, device, num_videos_per_prompt)

        # 3.2.2 Image latents.
```

sayakpaul replied:
I am okay with it but we're performing another resizing anyway within `_encode_image()`. I took the inspiration from SVD. Maybe it makes sense to move the center cropping and resizing out of the `_encode_image()` function to stay more consistent with the naming of the function? I'd be pretty okay with that.

yiyixuxu commented on src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py:
```diff
@@ -0,0 +1,880 @@
+
+        if not isinstance(image, torch.Tensor):
+            # https://github.com/ali-vilab/i2vgen-xl/blob/2539c9262ff8a2a22fa9daecbfd13f0a2dbc32d0/tools/inferences/inference_i2vgen_entrance.py#L114
+            image = _center_crop_wide(image, (width, width))
```
```suggestion
```

sayakpaul replied:
Sorry, why it's suggested to remove it? It is following the original implementation as commented.

yiyixuxu reviewed (commented):
looking great:)

sayakpaul commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,816 @@
+        context_embeddings = torch.cat([context_embeddings, encoder_hidden_states], dim=1)
+
+        image_latents_context_embeddings = _collapse_frames_into_batch(image_latents[:, :, :1, :])
+        image_latents_context_embeddings = self.image_latents_context_embedding(image_latents_context_embeddings)
```
I will let @DN6 address this if needed but generally, I agree.

sayakpaul commented:
@yiyixuxu I have updated the PR w.r.t your comments except for https://github.com/huggingface/diffusers/pull/6665#discussion_r1470834923. I will let @DN6 comment further here. 

Btw, @DN6 we need to generate checkpoints once again because the config has changed. It's better to have both FP32 and FP16 variants within the repo since the files are quite big!

DN6 commented:
@yiyixuxu Agreed that those embeddings can be differentiated better. Cool if I open up a follow up PR to clean it up?

sayakpaul commented:
> @yiyixuxu Agreed that those embeddings can be differentiated better. Cool if I open up a follow up PR to clean it up?

Which comment is being referred to herr?

patrickvonplaten commented:
Still a couple of failing tests here

sayakpaul commented:
@DN6 could you look into it once?

yiyixuxu commented:
> @yiyixuxu Agreed that those embeddings can be differentiated better. Cool if I open up a follow up PR to clean it up?

sure!

DN6 commented:
> Maybe we should rename the tensor to something like `image_context_emb` so that it does not have almost same name as the module layer :) a little bit hard to read like this for me. Just a suggestion, not super important though! :)
> 
> same goes with the `context_embeddings`/`context_embedding`

@sayakpaul This comment. But I addressed it and tests should also be fixed.

sayakpaul commented:
Cool thanks!

I think we still need to update the pipeline IDs here as well as the model card. We can then merge!

patrickvonplaten commented on src/diffusers/models/attention.py:
```diff
@@ -392,9 +393,9 @@ def forward(
+        if self.norm_type == "ada_norm_zero":
             ff_output = gate_mlp.unsqueeze(1) * ff_output
-        elif self.use_ada_layer_norm_single:
+        elif self.norm_type == "ada_norm_single":
```
Great job here! Much cleaner now :-)

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,679 @@
+        layers_per_block: int = 2,
+        norm_num_groups: Optional[int] = 32,
+        cross_attention_dim: int = 1024,
+        num_attention_heads: Optional[Union[int, Tuple[int]]] = 64,
```
nice clean up!

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,679 @@
+    ):
+        super().__init__()
+
+        self.sample_size = sample_size
```
should not be needed ideally because one can access it with `self.config.sample_size` but ok!

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,679 @@
+
+        self.transformer_in = TransformerTemporalModel(
+            num_attention_heads=8,
+            attention_head_dim=num_attention_heads,
```
```suggestion
            attention_head_dim=num_attention_heads,
```

attention_head_dim=num_attention_heads looks a bit weird - think the class TransformerTemporalModel has bad naming here no?

sayakpaul replied:
This came as a consequence of:

* https://github.com/huggingface/diffusers/pull/6665#discussion_r1470834759
* https://github.com/huggingface/diffusers/pull/6665#discussion_r1470835295
* https://github.com/huggingface/diffusers/pull/6665#discussion_r1470835388

See https://github.com/huggingface/diffusers/pull/6665#discussion_r1473695124 for more details.

yiyixuxu replied:
ohhh I saw where this is coming from. I think we should only keep `attention_head_dim` argument for `I2VGenXLUNet`, and then we can do 

```python
self.transformer_in = TransformerTemporalModel(
            num_attention_heads=8,
            attention_head_dim=attention_head_dim,
            in_channels=block_out_channels[0],
            num_layers=1,
            norm_num_groups=norm_num_groups,
        )
```

yiyixuxu replied:
cc @DN6 here

yiyixuxu replied:
`TransformerTemporalModel` does not have bad naming issue, and this is a new model so we definitely do not want to get the bad naming here ( here we force the user to pass what actually is `attention_head_dim` as `num_attention_head` is very much a bad naming practice )

sayakpaul replied:
I think we should keep both i.e., `attention_head_dim` and `num_attention_head`. Because if we only stick with `attention_head_dim`, we will end up with something like:

```python
down_block = get_down_block(
    down_block_type,
    num_layers=layers_per_block,
    in_channels=input_channel,
    out_channels=output_channel,
    temb_channels=time_embed_dim,
    add_downsample=not is_final_block,
    resnet_eps=1e-05,
    resnet_act_fn="silu",
    resnet_groups=norm_num_groups,
    cross_attention_dim=cross_attention_dim,
    num_attention_heads=attention_head_dim[i],
    downsample_padding=1,
    dual_cross_attention=False,
)
```

See how `num_attention_heads` is being assigned here. This again looks pretty bad naming-wise.

yiyixuxu replied:
see my comments here https://github.com/huggingface/diffusers/pull/6665/files#r1477715567

I think we can still do 
```
down_block = get_down_block(
    down_block_type,
    num_layers=layers_per_block,
    in_channels=input_channel,
    out_channels=output_channel,
    temb_channels=time_embed_dim,
    add_downsample=not is_final_block,
    resnet_eps=1e-05,
    resnet_act_fn="silu",
    resnet_groups=norm_num_groups,
    cross_attention_dim=cross_attention_dim,
    num_attention_heads=num_attention_heads[i],
    downsample_padding=1,
    dual_cross_attention=False,
)
```

patrickvonplaten commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,679 @@
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+def _to_tensor(inputs, device):
```
Let's please not forget to clean this up after we merge (cc @sayakpaul @yiyixuxu). 
This function is only used once - no need to move it into a function

yiyixuxu commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,679 @@
+        layers_per_block: int = 2,
+        norm_num_groups: Optional[int] = 32,
+        cross_attention_dim: int = 1024,
+        num_attention_heads: Optional[Union[int, Tuple[int]]] = 64,
```
this should be `attention_head_dim`, no? @DN6 @sayakpaul

sayakpaul replied:
I don't think so. I am unable to find out the comment from @patrickvonplaten, but look at how it's handled:

https://github.com/huggingface/diffusers/blob/c3369f56734beef9c4768d04cce490fdcc1c9162/src/diffusers/models/unets/unet_3d_condition.py#L130C1-L141C72

We can basically get rid of `attention_head_dim` and use `num_attention_heads` throughout to rectify the incorrect naming.

vladmandic commented:
this pr breaks a lot of existing pipelines as it redefines norm_types from what is very commonly used (and introduced a loong time ago).
for example, `StableDiffusionReferencePipeline` now fails with 
> AttributeError: 'BasicTransformerBlock' object has no attribute 'use_ada_layer_norm'

sayakpaul commented:
I welcome you to create an issue thread with a reproducible code snippet. We will fix it asap and if needed, do a patch release :-)

vladmandic commented:
i just did, i wanted to make a note here first.

yiyixuxu commented on src/diffusers/models/unets/unet_i2vgen_xl.py:
```diff
@@ -0,0 +1,679 @@
+                resnet_act_fn="silu",
+                resnet_groups=norm_num_groups,
+                cross_attention_dim=cross_attention_dim,
+                num_attention_heads=num_attention_heads[i],
```
so I think it is a mistake here 
`get_down_block()` expect `num_attention_heads` to be  correct `num_attention_heads`

 you look at how `num_attention_heads` is handled inside `UNet3DConditionModel`: 

First, it fix the "bad "model config name as soon as it receives the argument, by assigning `attention_head_dim` 
 to `num_attention_heads` here, so at this point, the variable`num_attention_heads` is "corrected" for the rest of the code https://github.com/huggingface/diffusers/blob/fdf55b1f1c25c7e2fedcdd0dc9c6f3db005453bb/src/diffusers/models/unets/unet_3d_condition.py#L141

from there, we can just pass it around as it is. e.g. it is passed as `num_attention_heads` in `get_down_block()`https://github.com/huggingface/diffusers/blob/fdf55b1f1c25c7e2fedcdd0dc9c6f3db005453bb/src/diffusers/models/unets/unet_3d_condition.py#L211


in our case, we:
1. also have the "bad" model config, i.e. for this model `num_attention_heads` is actually `attention_head_dim`
2. However, we only "correct" it for `TransformerTemporalModel` here. so for the rest of the code, `num_attention_heads` still means `attention_head_dim` - and that's incorrect 
```python
self.transformer_in = TransformerTemporalModel(
            num_attention_heads=8,
            attention_head_dim=attention_head_dim,
            in_channels=block_out_channels[0],
            num_layers=1,
            norm_num_groups=norm_num_groups,
        )
```

sayakpaul replied:
I am quite confused now. 

> First, it fix the "bad "model config name as soon as it receives the argument, by doing to num_attention_heads = attention_head_dim here, so at this point, the variablenum_attention_heads is "corrected" for the rest of the code

This is what we didn't want to do in this UNet i.e., the comments in https://github.com/huggingface/diffusers/pull/6665#discussion_r1473695124 suggested that there shouldn't be any `attention_head_dim`. We want to get rid of those corrections, made in `UNet3DConditionModel` for `num_attention_heads`. 

Do you mean we initialize another new variable called `attention_head_dim` based on what's passed to `num_attention_heads`?
